{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "icgc_mut = \"../data/datasets/PCAWG/mutations/snv_mnv_indel/final_consensus_passonly.snv_mnv_indel.icgc.public.maf\"\n",
    "tcgc_mut = \"../data/datasets/PCAWG/mutations/snv_mnv_indel/final_consensus_passonly.snv_mnv_indel.tcga.controlled.maf\"\n",
    "icgc_drivers = \"../data/datasets/PCAWG/driver_mutations/TableS3_panorama_driver_mutations_ICGC_samples.public.tsv\"\n",
    "tcga_drivers = \"../data/datasets/PCAWG/driver_mutations/TableS3_panorama_driver_mutations_TCGA_samples.controlled.tsv\"\n",
    "cnv_data = \"../data/datasets/PCAWG/mutations/cnv\"\n",
    "suppl_table = \"../data/datasets/PCAWG/supplementary Tables/Supplementary Table 1.csv\"\n",
    "mut_dir = \"../data/snv_mv_indels_by_cancer_subtype\"\n",
    "GENOME_WIDE_MUT_DATA = \"../data/genome_wide_mutation_data.tsv\"\n",
    "MANIFEST_FILE = \"../data/datasets/TCGA/gdc_manifest.2025-10-03.170810.txt\"\n",
    "SAMPLE_SHEET = \"../data/datasets/TCGA/gdc_sample_sheet.2025-10-03.tsv\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique samples in the whitelist: 2583\n"
     ]
    }
   ],
   "source": [
    "suppl_data = pd.read_csv(suppl_table, sep=\",\", header=0)\n",
    "whitelisted_samples = suppl_data[\"tumour_specimen_aliquot_id\"].unique().tolist()\n",
    "print(f\"Number of unique samples in the whitelist: {len(whitelisted_samples)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique TCGA samples in the whitelist: 801\n",
      "Number of unique TCGA patients in the whitelist: 801\n"
     ]
    }
   ],
   "source": [
    "tcga_specimen = suppl_data[suppl_data[\"submitted_specimen_id\"].str.contains(\"TCGA\")]\n",
    "tcga_specimen_ids = tcga_specimen[\"submitted_specimen_id\"].unique().tolist()\n",
    "print(f\"Number of unique TCGA samples in the whitelist: {len(tcga_specimen_ids)}\")\n",
    "tcga_patient_ids = [x.split(\"-\")[0] + \"-\" + x.split(\"-\")[1] + \"-\" + x.split(\"-\")[2] for x in tcga_specimen_ids]\n",
    "tcga_patient_ids = list(set(tcga_patient_ids))\n",
    "print(f\"Number of unique TCGA patients in the whitelist: {len(tcga_patient_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total TCGA patients in the manifest: 8536\n"
     ]
    }
   ],
   "source": [
    "manifest_df = pd.read_csv(MANIFEST_FILE, sep=\"\\t\")\n",
    "manifest_df.drop_duplicates(inplace=True)\n",
    "sample_df = pd.read_csv(SAMPLE_SHEET, sep=\"\\t\")\n",
    "sample_df.drop_duplicates(inplace=True)\n",
    "manifest_df = manifest_df.merge(sample_df, left_on=\"id\", right_on=\"File ID\", how=\"left\")\n",
    "manifest_df[\"Patient_ID\"] = manifest_df[\"Case ID\"].str.split(\", \").str[0]\n",
    "patients = manifest_df[\"Patient_ID\"].unique().tolist()\n",
    "print(\"Total TCGA patients in the manifest:\", len(patients))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of extra TCGA patients not in the PCAWG dataset: 7958\n"
     ]
    }
   ],
   "source": [
    "extra_tcga_patients = set(patients) - set(tcga_patient_ids)\n",
    "print(\"Number of extra TCGA patients not in the PCAWG dataset:\", len(extra_tcga_patients))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "icgc_mut = pd.read_csv(icgc_mut, sep=\"\\t\", low_memory=False)\n",
    "icgc_mut = icgc_mut = icgc_mut[[\"Hugo_Symbol\", \"Chromosome\", \"Start_position\", \"End_position\", \"Strand\", \"Reference_Allele\", \"Tumor_Seq_Allele1\", \"Tumor_Seq_Allele2\", \"Variant_Classification\", \"Tumor_Sample_Barcode\", \"Project_Code\", \"Donor_ID\"]]\n",
    "tcga_mut = pd.read_csv(tcgc_mut, sep=\"\\t\", low_memory=False)\n",
    "tcga_mut = tcga_mut[[\"Hugo_Symbol\", \"Chromosome\", \"Start_position\", \"End_position\", \"Strand\", \"Reference_Allele\", \"Tumor_Seq_Allele1\", \"Tumor_Seq_Allele2\", \"Variant_Classification\", \"Tumor_Sample_Barcode\", \"Project_Code\"]]\n",
    "mut_df = pd.concat([icgc_mut, tcga_mut], axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "drivers_icgc = pd.read_csv(icgc_drivers, sep=\"\\t\", low_memory=False)\n",
    "drivers_tcga = pd.read_csv(tcga_drivers, sep=\"\\t\", low_memory=False)\n",
    "driver_mut = pd.concat([drivers_icgc, drivers_tcga], axis=0, ignore_index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Number of unique samples in the whitelist: 2583\n"
     ]
    }
   ],
   "source": [
    "all_samples = mut_df[[\"Tumor_Sample_Barcode\", \"Project_Code\"]].drop_duplicates()\n",
    "all_samples[\"whitelisted\"] = all_samples[\"Tumor_Sample_Barcode\"].apply(lambda x: True if x in whitelisted_samples else False)\n",
    "all_samples[\"tissue\"] = all_samples[\"Project_Code\"].apply(lambda x: x.split(\"-\")[0])\n",
    "all_samples.rename(columns={\n",
    "    \"Tumor_Sample_Barcode\": \"sample_id\",\n",
    "\t\"Project_Code\": \"ttype\"\n",
    "}, inplace=True)\n",
    "all_samples = all_samples[all_samples[\"whitelisted\"] == True]\n",
    "print(f\"Number of unique samples in the whitelist: {len(all_samples)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(2354, 3)\n",
      "(2583, 5)\n"
     ]
    }
   ],
   "source": [
    "driver_count = driver_mut.groupby([\"sample_id\", \"ttype\"]).size().reset_index(name=\"num_drivers\")\n",
    "print(driver_count.shape)\n",
    "driver_count = all_samples.merge(driver_count, how=\"left\", on=[\"sample_id\", \"ttype\"])\n",
    "driver_count[\"num_drivers\"] = driver_count[\"num_drivers\"].fillna(0)\n",
    "print(driver_count.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "has_driver\n",
      "1    2354\n",
      "0     229\n",
      "Name: count, dtype: int64\n",
      "Percentage of samples with driver mutations: 91.13%\n",
      "Mean driver count: 5.304297328687572\n"
     ]
    }
   ],
   "source": [
    "# percentage of samples with driver mutations\n",
    "driver_count[\"has_driver\"] = driver_count[\"num_drivers\"].apply(lambda x: 1 if x > 0 else 0)\n",
    "driver_count[\"has_driver\"] = driver_count[\"has_driver\"].astype(int)\n",
    "counts = driver_count[\"has_driver\"].value_counts()\n",
    "print(counts)\n",
    "print(f\"Percentage of samples with driver mutations: {counts[1] / counts.sum() * 100:.2f}%\")\n",
    "# average driver count per sample\n",
    "mean_driver_count = driver_count[\"num_drivers\"].mean()\n",
    "print(f\"Mean driver count: {mean_driver_count}\")\n",
    "# average driver count per tissue type\n",
    "mean_driver_count_by_tissue = driver_count.groupby(\"tissue\")[\"num_drivers\"].mean().reset_index()\n",
    "mean_driver_count_by_tissue.sort_values(\"num_drivers\", ascending=False, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(10932, 3)\n",
      "count    10932.000000\n",
      "mean         1.253293\n",
      "std          0.466775\n",
      "min          1.000000\n",
      "25%          1.000000\n",
      "50%          1.000000\n",
      "75%          1.000000\n",
      "max          7.000000\n",
      "Name: num_drivers, dtype: float64\n"
     ]
    }
   ],
   "source": [
    "driver_gene_count = driver_mut.groupby([\"sample_id\", \"gene\"]).size().reset_index(name=\"num_drivers\")\n",
    "print(driver_gene_count.shape)\n",
    "# stats on number of drivers\n",
    "print(driver_gene_count[\"num_drivers\"].describe())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "application/vnd.microsoft.datawrangler.viewer.v0+json": {
       "columns": [
        {
         "name": "num_drivers",
         "rawType": "int64",
         "type": "integer"
        },
        {
         "name": "count",
         "rawType": "int64",
         "type": "integer"
        }
       ],
       "ref": "0c56226b-1cd7-4582-8acf-2218d8c53428",
       "rows": [
        [
         "1",
         "8281"
        ],
        [
         "2",
         "2557"
        ],
        [
         "3",
         "80"
        ],
        [
         "4",
         "8"
        ],
        [
         "5",
         "3"
        ],
        [
         "6",
         "2"
        ],
        [
         "7",
         "1"
        ]
       ],
       "shape": {
        "columns": 1,
        "rows": 7
       }
      },
      "text/plain": [
       "num_drivers\n",
       "1    8281\n",
       "2    2557\n",
       "3      80\n",
       "4       8\n",
       "5       3\n",
       "6       2\n",
       "7       1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "driver_gene_count[\"num_drivers\"].value_counts()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "cancer-model",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.18"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
